### PR TITLE
ICU-22277 correct collation error return of uninitialized length value while previous code return error

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22277
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [ ] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/i18n/ucol_res.cpp [hunk 1]
```diff
@@ -392,6 +392,7 @@ CollationLoader::loadFromData(UErrorCode &errorCode) {
     // but that created undesirable dependencies.
     int32_t length;
     const uint8_t *inBytes = ures_getBinary(binary.getAlias(), &length, &errorCode);
+    if(U_FAILURE(errorCode)) { return nullptr; }
```
Comment (markusicu) [inline icu4c/source/i18n/ucol_res.cpp:395]: optional -- simpler:
```suggestion
    int32_t length = 0;
    const uint8_t *inBytes = ures_getBinary(binary.getAlias(), &length, &errorCode);
```

File: icu4c/source/test/intltest/regcoll.cpp [hunk 2]
```diff
@@ -1249,6 +1249,26 @@ void CollationRegressionTest::TestBeforeWithTooStrongAfter() {
     }
 }
 
+void CollationRegressionTest::TestICU22277() {
+    IcuTestErrorCode errorCode(*this, "TestICU22277");
+    UErrorCode status = U_ZERO_ERROR;
+
+    Collator* c = Collator::createInstance("JA-u-Co-priVatE-KANa", status);
+    if(c != nullptr || U_SUCCESS(status)) {
+      errcheckln(status, "Collator should failed with with %s but no MemorySanitizer: use-of-uninitialized-value error",
```
Comment (markusicu) [inline icu4c/source/test/intltest/regcoll.cpp:1258]:
```suggestion
      errcheckln(status, "Collator should have failed with MemorySanitizer: use-of-uninitialized-value error - %s",
```

File: icu4c/source/test/intltest/regcoll.cpp [hunk 3]
```diff
@@ -1249,6 +1249,26 @@ void CollationRegressionTest::TestBeforeWithTooStrongAfter() {
...
+                 u_errorName(status));
+      delete c;
+      return;
+    }
+    c = Collator::createInstance("hE-U-cO-pRIVate-UNihan", status);
+    if(c != nullptr || U_SUCCESS(status)) {
+      errcheckln(status, "Collator should failed with with %s but no MemorySanitizer: use-of-uninitialized-value error",
```
Comment (markusicu) [inline icu4c/source/test/intltest/regcoll.cpp:1265]: ditto

**Comment by FrankYFTang** [comment]
PTAL

**Review by markusicu — APPROVED** [review]
lgtm tnx

**Pull request merged**